src/api/app/models/staging/staging_project.rb

```diff
@@ -139,7 +136,8 @@ def build_state
       :acceptable
     end
 
-    def relevant_status_reports(checkables, status_reports)
+    def status_reports(checkables)
+      status_reports = Status::Report.where(checkable: checkables)
```

Comment:
Does this produce the same query than the association? I would expect that the association does a join instead of a where...

Anyway, nitpick ...

Reply:
before:
```
[995:2.67] Scoped order is ignored, it's forced to be batch order.
[995:2.67]   Status::Report Load (0.3ms)  SELECT  `status_reports`.* FROM `status_reports` INNER JOIN `repository_architectures` ON `status_reports`.`checkable_id` = `repository_architectures`.`id` INNER JOIN `repositories` ON `repository_architectures`.`repository_id` = `repositories`.`id` WHERE `repositories`.`db_project_id` = 2 AND `status_reports`.`checkable_type` = 'RepositoryArchitecture' AND `status_reports`.`uuid` = 'f27678e3dee24d97' ORDER BY `status_reports`.`id` ASC LIMIT 1000
[995:2.67]   RepositoryArchitecture Load (0.2ms)  SELECT  `repository_architectures`.* FROM `repository_architectures` WHERE `repository_architectures`.`id` = 1 LIMIT 1
[995:2.68]    (0.2ms)  SELECT `status_checks`.`name` FROM `status_checks` WHERE `status_checks`.`status_reports_id` = 1
[995:2.68]    (0.2ms)  SELECT `status_checks`.`name` FROM `status_checks` WHERE `status_checks`.`status_reports_id` = 2
[995:2.68]   Status::Check Exists (0.2ms)  SELECT  1 AS one FROM `status_checks` WHERE `status_checks`.`status_reports_id` IN (1, 2) AND `status_checks`.`state` = 'pending' LIMIT 1
[995:2.68]   Status::Check Exists (0.2ms)  SELECT  1 AS one FROM `status_checks` WHERE `status_checks`.`status_reports_id` IN (1, 2) AND `status_checks`.`state` IN ('error', 'failure') LIMIT 1
```

after:

```
[1021:2.65]   Status::Report Load (0.4ms)  SELECT  `status_reports`.* FROM `status_reports` WHERE `status_reports`.`checkable_type` = 'RepositoryArchitecture' AND `status_reports`.`checkable_id` IN (SELECT `repository_architectures`.`id` FROM `repository_architectures` INNER JOIN `repositories` ON `repository_architectures`.`repository_id` = `repositories`.`id` WHERE `repositories`.`db_project_id` = 2 ORDER BY position) AND `status_reports`.`uuid` = 'ce22e61c46fd75a4' ORDER BY `status_reports`.`id` ASC LIMIT 1000
[1021:2.65]   RepositoryArchitecture Load (0.2ms)  SELECT  `repository_architectures`.* FROM `repository_architectures` WHERE `repository_architectures`.`id` = 1 LIMIT 1
[1021:2.65]    (0.2ms)  SELECT `status_checks`.`name` FROM `status_checks` WHERE `status_checks`.`status_reports_id` = 1
[1021:2.65]    (0.2ms)  SELECT `status_checks`.`name` FROM `status_checks` WHERE `status_checks`.`status_reports_id` = 2
[1021:2.65]   Status::Check Exists (0.2ms)  SELECT  1 AS one FROM `status_checks` WHERE `status_checks`.`status_reports_id` IN (1, 2) AND `status_checks`.`state` = 'pending' LIMIT 1
[1021:2.65]   Status::Check Exists (0.2ms)  SELECT  1 AS one FROM `status_checks` WHERE `status_checks`.`status_reports_id` IN (1, 2) AND `status_checks`.`state` IN ('error', 'failure') LIMIT 1
```

Can you spot a difference?

BTW: I guess the N+1 queries for check name should worry you (this is missing_checks).

Reply:
I don't think we want to do anything about missing_checks. We would need to fetch all status reports and all status checks and then do some ruby algorithm

On the other hand, if we'd load the checks anyway, we could stop using checks.fails.exist?, but loop through the ruby objects already in memory.